extide/gradle/src/org/netbeans/modules/gradle/execute/GradleDaemonExecutor.java

```diff
@@ -439,41 +441,70 @@ private static String gradleExecutable() {
         return Utilities.isWindows() ? "bin\\gradle.bat" : "bin/gradle"; //NOI18N
     }
 
-    public final ExecutorTask createTask(ExecutorTask process) {
+    // TODO one of the two methods can likely go from the API
+    // setTask is called first and signals the runnable to start
+    // the started runnable requires (!) the gradleTask so it can't be created in createTask
+    @Override
+    public void setTask(ExecutorTask task) {
         assert gradleTask == null;
-        gradleTask = new GradleTask(process);
+        gradleTask = new GradleTask(task);
+        super.setTask(task);
+    }
+
+    @Override
+    public final ExecutorTask createTask(ExecutorTask process) {
+        assert gradleTask != null;
+        assert task == process;
         return gradleTask;
     }
 
+    // task which can finish early, like a CompletableFuture
     private static final class GradleTask extends ExecutorTask {
-        private final ExecutorTask delegate;
-        private Integer result;
 
+        private final ExecutorTask delegate;
+        private volatile Integer result;
+        // is 0 when wrapper or delegate finished
+        private final CountDownLatch doneSignal = new CountDownLatch(1);
+        
         GradleTask(ExecutorTask delegate) {
             super(() -> {});
             this.delegate = delegate;
+            this.delegate.addTaskListener(t -> doneSignal.countDown());
```

Comment:
In this case, if called before finish, the `result` will appear to be 0 == success. Is that OK ? When searching for `finish` calls, I've found  paths in `GradleDaemonExecutor::run` - the exception handlers, that do not call gradleTask.finish at all, so these all will reach this lambda and produce result 0 ?

Reply:
not sure if I understand the question correctly. If someone calls `result()` and nothing calls the fast-path `finish(int)` the delegate task will count down the latch once it completes and the blocking `result()` method will now unblock and return `delegate.result()`.

the second case is that `finish(int)` is called before the delegate task completes, this would also unblock all waiting threads which wait on the latch, but this time it returns the `result` field without waiting for the delegate task.

the second case seems to be the intended purpose of the task wrapper to speed up the time-to-result.

Reply:
>  I've found paths in GradleDaemonExecutor::run - the exception handlers, that do not call gradleTask.finish at all, so these all will reach this lambda and produce result 0 ?

@sdedic this was the case before too, no?
```
            if (result != null) {
                return result;
            }
            return this.delegate.result();
```
this would have blocked on the delegate task and return the result. I am not 100% sure what you mean by 0.